dnadoh/cache.py

```diff
@@ -0,0 +1,155 @@
+    `staticmethod`.  # because depends on the cache path?
+"""
+
+import csv
+import filecmp
+import requests
+import shutil
+import validators
+
+from pathlib import Path
+from urllib.parse import urljoin
+
+
+CONFIG = {
+    "cache": Path(__file__).parent.parent.joinpath("cache"),  # prints DNA-doh rootdir + cache
+    "remote": Path(__file__).parent.parent.joinpath("remote"),
+}
+
+DOWNLOAD_LIMIT = 2.5e8
+
+CACHE_INDEX = "cache_index.csv"
+CACHE_HEADER = ["cache_path", "remote_path"]
+
+class FileCache:
+    """Cache large remote files."""
+
+    def __init__(self, cache_dir=None, remote_url=None):
+        if cache_dir:
+            self.cache_dir = cache_dir
+        else:
+            self.cache_dir = CONFIG["cache"]
+        
+        if remote_url:
+            self.remote_url = remote_url
+        else:
+            self.remote_url = CONFIG["remote"]
+        
+        self.cache_index = Path(self.cache_dir, CACHE_INDEX)
+        self.setup_cache()
+
+        
+    def get(self, requested_file):
+        """Return path to cached copy of file, getting as needed."""
+
+        # special handling of URLs
+        if validators.url(requested_file):
+            # extact filename from URL as that'll be how it's saved in the cache that we need to check
+            extracted_filename = Path(requested_file).name
+            cache_path = Path(self.cache_dir, extracted_filename)
+            remote_path = requested_file
+        else:
+            cache_path = Path(self.cache_dir, requested_file)
+            remote_path = Path(self.remote_url, requested_file)
+
+        # first check if file is in the cache
+        if cache_path.exists():
+            # confirm that it's in the cache index and if not then add it
+            if not self.check_if_in_cache(cache_path):
+                self.add_to_cache_index(remote_path, cache_path)
+        else:
+            if validators.url(requested_file):
+                self._download_file(requested_file, cache_path)
+            else:
+                self.handle_local_file(requested_file, cache_path)
+        
+        return str(cache_path)
+
+    def clear(self):
+        """Clear the cache, deleting local files."""
+        for filename in self.cache_dir.iterdir():
+            filename.unlink()
+        self.setup_cache()
+
+    def _get_file(self, remote_path, local_path):
+        """Simulate getting a remote file."""
+        remote_path = Path(self.remote_url, remote_path)
+        # print(remote_path)
+        if not remote_path.exists():
+            raise ValueError(f"{remote_path} does not exist")
+        shutil.copyfile(remote_path, local_path)
+
+    def _download_file(self, remote_url, local_path):
+        """Downloads a file from a remote url"""
+        h = requests.head(remote_url, allow_redirects=True)
+        header = h.headers
+        content_length = header.get("content-length", None)
+        # i think content length only works for txt files? couldn't see one with an image
```

Comment:
it should be there for all files: we can debug together

Reply:
figured out that when I was trying to access an image on github (e.g.,: https://github.com/tammy-dg/tammys-simple-web-server/blob/master/test/test.png), need to add `?raw=true` at the end of the URL and then it will have a `content-length` key. Still unknown as to why that's not in the original URL though.

Reply:
Greg to follow up